relayer/src/link/relay_path.rs

```diff
@@ -1438,62 +1598,90 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
     /// waiting for the delay period to pass.
-    /// Returns `None` if there is no operational data scheduled.
-    pub(crate) fn fetch_scheduled_operational_data(&self) -> Option<OperationalData> {
-        let odata = self
-            .src_operational_data
-            .pop_front()
-            .or_else(|| self.dst_operational_data.pop_front());
+    /// Returns `Ok(None)` if there is no operational data scheduled.
+    pub(crate) fn fetch_scheduled_operational_data(
+        &self,
+    ) -> Result<Option<OperationalData>, LinkError> {
+        let (time_left, blocks_left, odata) =
+            if let Some(odata) = self.src_operational_data.pop_front() {
+                let src_chain_time = self.src_time_latest()?;
+                let src_block_time = self.src_max_block_time()?;
+                let src_latest_height = self.src_latest_height()?;
+                (
+                    odata.conn_time_delay_elapsed(src_chain_time).err(),
+                    odata
+                        .conn_block_delay_elapsed(src_block_time, src_latest_height)
+                        .err(),
+                    Some(odata),
+                )
+            } else if let Some(odata) = self.dst_operational_data.pop_front() {
+                let dst_chain_time = self.dst_time_latest()?;
+                let dst_block_time = self.dst_max_block_time()?;
+                let dst_latest_height = self.dst_latest_height()?;
+                (
+                    odata.conn_time_delay_elapsed(dst_chain_time).err(),
+                    odata
+                        .conn_block_delay_elapsed(dst_block_time, dst_latest_height)
+                        .err(),
+                    Some(odata),
+                )
+            } else {
+                (None, None, None)
+            };
 
         if let Some(odata) = odata {
-            // Check if the delay period did not completely elapse
-            let delay_left = self
-                .channel
-                .connection_delay
-                .checked_sub(odata.scheduled_time.elapsed());
-
-            match delay_left {
-                None => info!(
+            match (time_left, blocks_left) {
+                (None, None) => info!(
                     "ready to fetch a scheduled op. data with batch of size {} targeting {}",
                     odata.batch.len(),
                     odata.target,
                 ),
-                Some(delay_left) => {
+                (Some(delay_time), _) => {
                     info!(
                         "waiting ({:?} left) for a scheduled op. data with batch of size {} targeting {}",
-                        delay_left,
+                        delay_time,
                         odata.batch.len(),
                         odata.target,
                     );
 
                     // Wait until the delay period passes
-                    thread::sleep(delay_left);
+                    thread::sleep(delay_time);
```

Comment:
Should we add some warning message to the user if `delay_time != 0` so it is clear that the CLI is not broken but needs to wait for the delay_time?

Reply:
I noticed that there is an `info!()` log [above that code](https://github.com/informalsystems/ibc-rs/pull/1917/files#diff-2a5a700c0b1975ad9d64941997aa53e0b6e26e577fa8f1e9a2aba4abe7cf1f95R1664-R1665), do you mean we should change it to `warn!()`? Or should we add a CLI output line to the packet CLI commands?

Reply:
Ah, didn't realize we have an `info`. I think it's fine.